internal: implement `--profiler:on` support with MIR pass

zerbina opened the pull request:
## Summary

Make routine instrumentation with `nimProfile` calls a MIR pass, moving
more logic out of the C code generator and working towards
`--profiler:on` support with all backends. `nimProfile` calls are also
no longer inserted into loops within `.asmNoStackFrame` routines.

## Details

* require a `ModuleGraph` instance for `mirpasses.applyPasses`, so that
  compilerprocs can be looked up
* add the `injectProfilerCalls` MIR pass. It injects `nimProfile` in
  the same way that `cgen` does
* remove injection of `nimProfile` calls from `cgen`
* don't instrument loops within `.asmNoStackFrame` procedures;
  injection of `nimProfile` at procedure entry was already skipped for
  these procedures
* temporarily remove the `optProfiler` option from symbols when in JIT
  mode, that code generated for compile-time execution is not
  instrumented
* add a test for the `--profiler:on` feature, to make sure it works and
  keeps working

zerbina committed:
mirpasses: inject `nimProfile` calls with a MIR pass

The pass injects the calls at same places that `cgen` injects them,
with the difference that no `nimProfile` calls are injected into loops
within `.asmNoStackFrame` (pure) procedures.

To be able to query the `.compilerProc`, `applyPasses` needs access to
`ModuleGraph`; the unused `config` parameter is replaced with a
`ModuleGraph`.

zerbina committed:
cgen: remove injection of `nimProfile` calls

The MIR pass handles the injection now.

zerbina committed:
tests: add a test for the `--profiler` option

It doubles as a small tutorial on how to makes use of it (without
using `nimprof`).

zerbina committed:
cgen: remove now-unused `appcg` overload

saem commented on compiler/mir/mirpasses.nim:
```diff
@@ -455,6 +488,16 @@ proc applyPasses*(body: var MirBody, prc: PSym, env: var MirEnv,
+  # optimization
+  if (sfPure notin prc.flags) and (optProfiler in prc.options) and
+     (target in {targetC, targetJs}):
+    # XXX: not enabled for the VM because compile-time and run-time cannot
```
Could have the compiler pass system temporarily remove the flag from options and then restore it prior to the code generation pass.

zerbina replied:
Hm, no, I haven't considered doing this, but it would work, and the implementation cost is small. I'll implement it.

saem commented on tests/compilerfeatures/tprofiler.nim:
```diff
@@ -0,0 +1,67 @@
+# nothing will happen before the hook is set
+profilerHook = profileCallback
+# the "profiling requested" callback guards whether to invoke the profile
+# callback is invoked
```
```suggestion
# callback
```

saem reviewed (approved):
Minor typo you may want to fix, otherwise the suggestion for limiting the profile during compile time is very soft. You might have already thought of it yourself.

zerbina committed:
mirpasses: remove VM target special casing

`vmjit` now temporarily excludes the `optProfiler` option from the
symbol, allowing `applyPasses` to ignore the selected target.

The `tprofiler` test still doesn't work, however.

zerbina committed:
fix a typo

zerbina commented:
/merge

zerbina commented:
/merge

zerbina commented:
/merge

Aaand another try. The first failure was due to the flaky `tfragment_alloc.nim` test, the second one due to an unusual MacOS failure where xcodebuild could not be located.
